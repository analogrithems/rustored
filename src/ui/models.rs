use aws_sdk_s3::primitives::DateTime as AwsDateTime;
use std::fmt;
use std::fmt::Debug;

/// Configuration for S3 connection
#[derive(Clone, Debug)]
pub struct S3Config {
    pub bucket: String,
    pub region: String,
    pub prefix: String,
    pub endpoint_url: String,
    pub access_key_id: String,
    pub secret_access_key: String,
    pub path_style: bool,
    pub error_message: Option<String>,
}

impl S3Config {
    pub fn mask_secret(&self, secret: &str) -> String {
        if secret.len() <= 4 {
            return "*".repeat(secret.len());
        }
        let visible_chars = 4;
        let hidden_chars = secret.len() - visible_chars;
        format!("{}{}", "*".repeat(hidden_chars), &secret[hidden_chars..])
    }

    pub fn masked_access_key(&self) -> String {
        self.mask_secret(&self.access_key_id)
    }

    pub fn masked_secret_key(&self) -> String {
        self.mask_secret(&self.secret_access_key)
    }
}

/// Configuration for PostgreSQL connection
#[derive(Clone, Debug, PartialEq, Default)]
pub enum RestoreTarget {
    #[default]
    Postgres,
    Elasticsearch,
    Qdrant,
}

#[derive(Debug, Clone)]
pub struct PostgresConfig {
    pub host: Option<String>,
    pub port: Option<u16>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub use_ssl: bool,
    pub db_name: Option<String>,
}

/// Input mode for the UI
#[derive(Debug, PartialEq)]
pub enum InputMode {
    Normal,
    Editing,
}

/// Metadata for a backup
#[derive(Clone, Debug, PartialEq)]
pub struct BackupMetadata {
    pub key: String,
    pub size: i64,
    pub last_modified: AwsDateTime,
}

/// State of the popup
#[derive(Debug, PartialEq)]
pub enum PopupState {
    Hidden,
    ConfirmRestore(BackupMetadata),
    Downloading(BackupMetadata, f32, f64),
    ConfirmCancel(BackupMetadata, f32, f64),
    Restoring(BackupMetadata, f32),  // Snapshot being restored, progress percentage
    TestS3Result(String),
    TestPgResult(String),
    Error(String),
    Success(String),
}

/// Focus field for the UI
#[derive(Debug, PartialEq, Clone, Copy)]
pub enum FocusField {
    // S3 Settings (10-19)
    Bucket,          // Alt+1
    Region,          // Alt+2
    Prefix,          // Alt+3
    EndpointUrl,     // Alt+4
    AccessKeyId,     // Alt+5
    SecretAccessKey, // Alt+6
    PathStyle,       // Alt+7

    // PostgreSQL Settings (20-29)
    PgHost,          // Alt+q
    PgPort,          // Alt+w
    PgUsername,      // Alt+e
    PgPassword,      // Alt+r
    PgSsl,          // Alt+t
    PgDbName,        // Alt+y
    SnapshotList,
    RestoreTarget,
    EsHost,
    EsIndex,
    QdrantApiKey,
}

impl fmt::Display for FocusField {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FocusField::Bucket => write!(f, "Bucket"),
            FocusField::Region => write!(f, "Region"),
            FocusField::Prefix => write!(f, "Prefix"),
            FocusField::EndpointUrl => write!(f, "Endpoint URL"),
            FocusField::AccessKeyId => write!(f, "Access Key ID"),
            FocusField::SecretAccessKey => write!(f, "Secret Access Key"),
            FocusField::PathStyle => write!(f, "Path Style"),
            FocusField::PgHost => write!(f, "PostgreSQL Host"),
            FocusField::PgPort => write!(f, "PostgreSQL Port"),
            FocusField::PgUsername => write!(f, "PostgreSQL Username"),
            FocusField::PgPassword => write!(f, "PostgreSQL Password"),
            FocusField::PgSsl => write!(f, "PostgreSQL SSL"),
            FocusField::PgDbName => write!(f, "PostgreSQL Database"),
            FocusField::SnapshotList => write!(f, "Snapshot List"),
        FocusField::RestoreTarget => write!(f, "Restore Target"),
        FocusField::EsHost => write!(f, "Elasticsearch/Qdrant Host"),
        FocusField::EsIndex => write!(f, "Index/Collection"),
        FocusField::QdrantApiKey => write!(f, "Qdrant API Key"),
        }
    }
}
