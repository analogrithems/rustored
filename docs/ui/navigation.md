# Rustored UI Navigation Guide

This document provides a comprehensive guide to the navigation system in the Rustored terminal user interface (TUI).

## UI Layout

The Rustored UI is organized according to TDD rule #7, with a two-row layout:

```ascii
┌─────────────────┬─────────────────┬─────────────────┐
│                 │                 │                 │
│   S3 Settings   │ Restore Target  │ Target-Specific │
│                 │   Selection     │    Settings     │
│                 │                 │                 │
├─────────────────┴─────────────────┴─────────────────┤
│                                                     │
│                 Snapshot Browser                    │
│                                                     │
└─────────────────────────────────────────────────────┘
```

1. **Top Row**: Contains three panels side by side:
   - **S3 Settings** (left panel): Configuration for S3 connection parameters
   - **Restore Target Selection** (middle panel): Options to select the target datastore type (PostgreSQL, Elasticsearch, Qdrant)
   - **Target-Specific Settings** (right panel): Dynamic settings panel that changes based on the selected restore target

2. **Bottom Row**: Contains the snapshot browser that displays available snapshots from S3

For a more detailed view of the UI layout, see the [full diagram](../images/ui_layout.svg).

## Keyboard Navigation

### Window Navigation

| Key | Action |
|-----|--------|
| `Tab` | Cycle between main window sections (S3 Settings → Restore Target Settings → Snapshot List → S3 Settings) |
| `q` | Quit the application |
| `Ctrl+Z` | Suspend the application (Unix systems only) |
| `r` | Reload snapshots from S3 |
| `t` | When focus is on S3 Settings: Test S3 connection |

### Restore Target Selection

| Key | Action |
|-----|--------|
| `1` | Select PostgreSQL as the restore target |
| `2` | Select Elasticsearch as the restore target |
| `3` | Select Qdrant as the restore target |

When you select a different restore target, the Restore Settings panel will automatically update to show the appropriate settings for that target. Additionally, the focus will move to the first field in the selected target's settings if it wasn't already on a field for that target.

### Within-Window Navigation

| Key | Action |
|-----|--------|
| `↓` or `j` | Move to the next field within the current window |
| `↑` or `k` | Move to the previous field within the current window |
| `Enter` | When in Snapshot List: Select the highlighted snapshot for restore |

The up/down navigation wraps around within each window section. For example, if you're on the last field in the S3 Settings window and press Down, you'll cycle back to the first field in that window.

### Field Editing

| Key | Action |
|-----|--------|
| `Enter` | When on a field: Enter edit mode for the currently focused field |
| `Enter` | When already in edit mode: Save the edited value |
| `Esc` | Cancel editing |

## Focus Indicators

The currently focused field or section is highlighted in yellow. This provides visual feedback about which element will receive keyboard input.

## Popup Windows

Popup windows appear in certain situations:

- **Confirmation**: When confirming an action like restoring a snapshot
- **Progress**: When downloading or restoring a snapshot
- **Error/Success**: When an operation completes or fails

Press `Esc` or `Enter` to dismiss most popups.

## Example Workflow

1. Use `Tab` to navigate to the S3 Settings window
2. Use `↓`/`↑` or `j`/`k` to navigate between fields within the S3 Settings window
3. Press `Enter` to edit the currently focused field
4. Press `t` to test the S3 connection
5. Press `r` to load snapshots from S3
6. Press `Tab` to navigate to the Restore Target Settings window
7. Press `1`, `2`, or `3` to select the desired restore target
8. Use `↓`/`↑` or `j`/`k` to navigate between fields within the Restore Target Settings window
9. Press `Tab` to navigate to the Snapshot List window
10. Use `↓`/`↑` or `j`/`k` to select a snapshot
11. Press `Enter` to confirm and start the restore process

## Implementation Details

The navigation system is implemented in the `RustoredApp.handle_key_event` method, which processes all key events and updates the application state accordingly. The UI rendering is handled by the `renderer.rs` file, which uses the Ratatui library to create the layout and render the components.

When switching between restore targets, the focus is automatically updated to ensure a smooth user experience. This is achieved through the following code in the key event handler:

```rust
// Tab key handling - only changes window focus
KeyCode::Tab => {
    // Cycle between main window sections only
    self.focus = match self.focus {
        // S3 Settings fields - move to Restore Target settings
        FocusField::Bucket | FocusField::Region | /* other S3 fields */ => {
            // Move to restore target settings
            match self.restore_target {
                RestoreTarget::Postgres => FocusField::PgHost,
                RestoreTarget::Elasticsearch => FocusField::EsHost,
                RestoreTarget::Qdrant => FocusField::EsHost,
            }
        }
        // Restore Target settings - move to Snapshot List
        FocusField::PgHost | /* other Postgres fields */ |
        FocusField::EsHost | /* other ES fields */ => FocusField::SnapshotList,
        // Snapshot list - move back to S3 Settings
        FocusField::SnapshotList => FocusField::Bucket,
        // Default case
        _ => FocusField::Bucket,
    };
}

// Down key handling - navigates within a window
KeyCode::Down | KeyCode::Char('j') => {
    if self.focus == FocusField::SnapshotList {
        // Navigate snapshot list
        self.snapshot_browser.selected_index = 
            (self.snapshot_browser.selected_index + 1) % self.snapshot_browser.snapshots.len();
    } else {
        // Navigate within settings windows
        self.focus = match self.focus {
            // S3 Settings navigation
            FocusField::Bucket => FocusField::Region,
            // ... other field navigation logic
        };
    }
}
