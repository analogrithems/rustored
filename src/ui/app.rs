use std::io::stdout;
use log::debug;
use std::time::Duration;
use anyhow::Result;
use crossterm::event::{self, Event, KeyCode, KeyModifiers};
use crossterm::terminal::{disable_raw_mode, enable_raw_mode};
use crossterm::execute;
use crossterm::terminal::{EnterAlternateScreen, LeaveAlternateScreen};
use crossterm::event::DisableMouseCapture;
use crossterm::event::EnableMouseCapture;
use libc::SIGTSTP;
use libc::raise;
use ratatui::backend::Backend;
use ratatui::Terminal;

use crate::ui::models::InputMode;
use crate::ui::rustored::RustoredApp;

/// Run the TUI application, delegating to RustoredApp
pub async fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut RustoredApp) -> Result<Option<String>> {
    // Initial load of snapshots
    if let Err(e) = app.snapshot_browser.load_snapshots().await {
        debug!("Failed to load snapshots: {}", e);
    }

    loop {
        // Draw UI
        terminal.draw(|f| crate::ui::renderer::ui::<B>(f, app))?;

        // Handle events
        if event::poll(Duration::from_millis(100))? {
            if let Event::Key(key) = event::read()? {
                // Handle suspend on Ctrl+Z when in Normal mode
                if app.input_mode == InputMode::Normal && key.modifiers.contains(KeyModifiers::CONTROL) && key.code == KeyCode::Char('z') {
                    // Exit TUI mode
                    disable_raw_mode()?;
                    execute!(stdout(), LeaveAlternateScreen, DisableMouseCapture)?;
                    // Suspend process
                    unsafe { raise(SIGTSTP); }
                    // On resume, re-enter TUI
                    enable_raw_mode()?;
                    execute!(stdout(), EnterAlternateScreen, EnableMouseCapture)?;
                    continue;
                }
                
                // Pass the key event to the app
                if let Some(snapshot_path) = app.handle_key_event::<B>(key).await? {
                    return Ok(Some(snapshot_path));
                }
            }
        }
    }
}
