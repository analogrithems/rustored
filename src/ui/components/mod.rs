pub mod s3_settings;
pub mod restore_target;
pub mod snapshot_list;
pub mod popups;

// Re-export all components for easier imports
pub use s3_settings::*;
pub use restore_target::*;
pub use snapshot_list::*;
pub use popups::*;
