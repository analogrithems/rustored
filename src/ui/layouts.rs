use ratatui::layout::{Constraint, Direction, Layout, Rect};
use log::debug;

/// Helper function to create a centered rect
pub fn centered_rect(percent_x: u16, height: u16, r: Rect) -> Rect {
    debug!("Creating centered rect with percent_x: {}, height: {}, in area: {:?}", percent_x, height, r);
    let popup_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(
            [
                Constraint::Percentage((100 - height) / 2),
                Constraint::Length(height),
                Constraint::Percentage((100 - height) / 2),
            ]
            .as_ref(),
        )
        .split(r);

    let result = Layout::default()
        .direction(Direction::Horizontal)
        .constraints(
            [
                Constraint::Percentage((100 - percent_x) / 2),
                Constraint::Percentage(percent_x),
                Constraint::Percentage((100 - percent_x) / 2),
            ]
            .as_ref(),
        )
        .split(popup_layout[1])[1];
    
    debug!("Created centered rect: {:?}", result);
    result
}
