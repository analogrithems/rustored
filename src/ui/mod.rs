// UI module for the postgres manager
pub mod models;
pub mod browser;
pub mod renderer;

// Re-export the main entry point
pub use browser::run_tui;
