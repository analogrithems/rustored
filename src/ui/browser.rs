use anyhow::{Result, anyhow};
use aws_sdk_s3::operation::list_objects_v2::ListObjectsV2Output;
use aws_sdk_s3::operation::get_object::GetObjectOutput;
use aws_sdk_s3::Client as S3Client;

use log::{debug, warn};
use tokio::io::AsyncReadExt;
use std::path::Path;
use std::fs::File;
use std::io::Write;
use crate::ui::models::{S3Config, PopupState, FocusField, BackupMetadata};
use chrono::Utc;

/// Component for S3 snapshot browsing
pub struct SnapshotBrowser {
    // S3 Configuration
    pub s3_config: S3Config,
    
    // S3 Client
    s3_client: Option<S3Client>,
    
    // UI State
    pub focus: FocusField,
    pub input_mode: crate::ui::models::InputMode,
    pub input_buffer: String,
    pub snapshots: Vec<BackupMetadata>,
    pub selected_index: usize,
    pub popup_state: PopupState,
}

impl SnapshotBrowser {
    /// Create a new SnapshotBrowser with default settings
    pub fn new(s3_config: S3Config) -> Self {
        Self {
            s3_config,
            s3_client: None,
            focus: FocusField::SnapshotList,
            input_mode: crate::ui::models::InputMode::Normal,
            input_buffer: String::new(),
            snapshots: Vec::new(),
            selected_index: 0,
            popup_state: PopupState::Hidden,
        }
    }

    /// Initialize the S3 client based on current settings
    pub async fn init_client(&mut self) -> Result<()> {
        self.s3_client = Some(self.s3_config.create_client()?);
        Ok(())
    }

    /// Load snapshots from S3
    pub async fn load_snapshots(&mut self) -> Result<()> {
        debug!("Loading snapshots from S3");
        
        // Initialize client if needed
        if self.s3_client.is_none() {
            debug!("S3 client not initialized, initializing now");
            self.init_client().await?;
        }
        
        if let Some(client) = &self.s3_client {
            debug!("Using S3 client to list objects in bucket: {}", self.s3_config.bucket);
            
            let list_objects_result = client
                .list_objects_v2()
                .bucket(&self.s3_config.bucket)
                .prefix(&self.s3_config.prefix)
                .send()
                .await;
                
            match list_objects_result {
                Ok(output) => {
                    debug!("Successfully listed objects, parsing results");
                    self.parse_s3_output(output);
                    Ok(())
                }
                Err(e) => {
                    debug!("Failed to list objects: {}", e);
                    Err(anyhow!("Failed to list objects: {}", e))
                }
            }
        } else {
            debug!("S3 client not available");
            Err(anyhow!("S3 client not initialized"))
        }
    }

    /// Parse S3 output and populate snapshots
    fn parse_s3_output(&mut self, output: ListObjectsV2Output) {
        self.snapshots.clear();
        self.selected_index = 0;

        if let Some(contents) = output.contents {
            for obj in contents {
                // Skip directory-like objects (ones that end with /)
                if let Some(key) = &obj.key {
                    if key.ends_with('/') {
                        continue;
                    }
                    
                    // Only include objects that match our prefix and have content
                    if key.starts_with(&self.s3_config.prefix) {
                        let metadata = BackupMetadata {
                            key: key.clone(),
                            size: obj.size.unwrap_or(0),
                            last_modified: obj.last_modified
                                .map(|dt| dt.as_secs_f64())
                                .unwrap_or_else(|| Utc::now().timestamp() as f64),
                        };
                        self.snapshots.push(metadata);
                    }
                }
            }
        }
        
        // Sort by most recent first
        self.snapshots.sort_by(|a, b| b.last_modified.partial_cmp(&a.last_modified).unwrap_or(std::cmp::Ordering::Equal));
        
        debug!("Loaded {} snapshots", self.snapshots.len());
    }

    /// Download snapshot to a local file
    pub async fn download_snapshot(
        &mut self,
        snapshot: &BackupMetadata,
        tmp_path: &Path,
    ) -> Result<Option<String>> {
        if let Some(client) = &self.s3_client {
            self.popup_state = PopupState::Downloading(snapshot.clone(), 0.0, 0.0);
            
            // Set popup state for download
            self.popup_state = PopupState::Downloading(snapshot.clone(), 0.0, 0.0);
            
            match client
                .get_object()
                .bucket(&self.s3_config.bucket)
                .key(&snapshot.key)
                .send()
                .await
            {
                Ok(output) => {
                    debug!("Download started for {}", snapshot.key);
                    
                    // Ensure parent directory exists
                    if let Some(parent) = tmp_path.parent() {
                        std::fs::create_dir_all(parent)?;
                    }
                    
                    // Save to file
                    if let Ok(file) = File::create(tmp_path) {
                        let tmp_path_str = tmp_path.to_string_lossy().to_string();
                        if let Err(e) = self.save_stream_to_file(output, snapshot, file).await {
                            warn!("Error saving file: {}", e);
                            self.popup_state = PopupState::Error(format!("Download failed: {}", e));
                            // Set error popup state
                            return Ok(None);
                        }
                        return Ok(Some(tmp_path_str));
                    } else {
                        debug!("Could not create file at {:?}", tmp_path);
                        return Ok(None);
                    }
                }
                Err(e) => {
                    debug!("Failed to download snapshot {}: {}", snapshot.key, e);
                    return Ok(None);
                }
            }
        } else {
            debug!("Download attempted but S3 client not initialized");
            return Ok(None);
        }
    }

    /// Save stream to file with progress updates
    async fn save_stream_to_file(
        &mut self,
        output: GetObjectOutput,
        snapshot: &BackupMetadata,
        mut file: File,
    ) -> Result<()> {
        let mut body = output.body.into_async_read();
        let size = snapshot.size as f64;
        let mut downloaded: usize = 0;
        let mut buffer = [0; 1024 * 64]; // 64KB buffer
        let start_time = std::time::Instant::now();
        let mut last_update = std::time::Instant::now();
        
        loop {
            match body.read(&mut buffer).await {
                Ok(0) => break, // EOF
                Ok(n) => {
                    // Write to file
                    file.write_all(&buffer[0..n])?;
                    downloaded += n;
                    
                    // Update progress at most 10 times per second
                    let now = std::time::Instant::now();
                    if now.duration_since(last_update).as_millis() > 100 {
                        last_update = now;
                        let elapsed = now.duration_since(start_time).as_secs_f64();
                        let rate = if elapsed > 0.0 { downloaded as f64 / elapsed } else { 0.0 };
                        let progress = downloaded as f64 / size;
                        
                        // Update popup state
                        self.popup_state = PopupState::Downloading(snapshot.clone(), progress as f32, rate);
                        
                        // Check for user cancel
                        if let PopupState::ConfirmCancel(_, _, _) = self.popup_state {
                            return Err(anyhow!("Download cancelled by user"));
                        }
                    }
                }
                Err(e) => return Err(anyhow!("Error reading from S3: {}", e)),
            }
        }
        
        debug!("Download complete: {}", snapshot.key);
        self.popup_state = PopupState::Success(format!("Download complete: {}", snapshot.key));
        
        Ok(())
    }

    // The restore_snapshot method has been moved to RustoredApp

    // Key handling has been moved to RustoredApp
}
