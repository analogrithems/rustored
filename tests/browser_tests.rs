use insta::assert_debug_snapshot;

// We need to use the crate name directly since we're in an integration test
use rustored::ui::models::{BackupMetadata, FocusField, InputMode, PopupState, PostgresConfig, S3Config};
use aws_sdk_s3::primitives::DateTime as AwsDateTime;

// Helper function to create a test AwsDateTime with a fixed timestamp
fn create_test_aws_datetime() -> AwsDateTime {
    // Use a fixed timestamp for testing (2023-01-01 12:00:00 UTC)
    AwsDateTime::from_secs(1672574400)
}

#[test]
fn test_s3_config() {
    let s3_config = S3Config {
        bucket: "test-bucket".to_string(),
        region: "us-west-2".to_string(),
        prefix: "test-prefix".to_string(),
        endpoint_url: "https://test-endpoint.com".to_string(),
        access_key_id: "test-access-key".to_string(),
        secret_access_key: "test-secret-key".to_string(),
        path_style: false,
        error_message: None,
    };

    assert_debug_snapshot!(s3_config);
}

#[test]
fn test_postgres_config() {
    let pg_config = PostgresConfig {
        host: Some("localhost".to_string()),
        port: Some(5432),
        username: Some("postgres".to_string()),
        password: Some("password".to_string()),
        use_ssl: false,
        db_name: Some("postgres".to_string()),
    };

    assert_debug_snapshot!(pg_config);
}

#[test]
fn test_backup_metadata() {
    let datetime = create_test_aws_datetime();
    let backup = BackupMetadata {
        key: "test-snapshot-1.sql.gz".to_string(),
        size: 1024 * 1024 * 10, // 10 MB
        last_modified: datetime,
    };

    assert_debug_snapshot!(backup);
}

#[test]
fn test_popup_states() {
    let datetime = create_test_aws_datetime();
    let backup = BackupMetadata {
        key: "test-snapshot-1.sql.gz".to_string(),
        size: 1024 * 1024 * 10, // 10 MB
        last_modified: datetime,
    };

    let hidden = PopupState::Hidden;
    let confirm_restore = PopupState::ConfirmRestore(backup.clone());
    let downloading = PopupState::Downloading(backup.clone(), 0.5, 1024.0 * 1024.0);
    let confirm_cancel = PopupState::ConfirmCancel(backup.clone(), 0.5, 1024.0 * 1024.0);
    let error = PopupState::Error("Test error message".to_string());
    let success = PopupState::Success("Test success message".to_string());

    assert_debug_snapshot!("popup_state_hidden", hidden);
    assert_debug_snapshot!("popup_state_confirm_restore", confirm_restore);
    assert_debug_snapshot!("popup_state_downloading", downloading);
    assert_debug_snapshot!("popup_state_confirm_cancel", confirm_cancel);
    assert_debug_snapshot!("popup_state_error", error);
    assert_debug_snapshot!("popup_state_success", success);
}

#[test]
fn test_focus_fields() {
    assert_debug_snapshot!("focus_field_snapshot_list", FocusField::SnapshotList);
    assert_debug_snapshot!("focus_field_bucket", FocusField::Bucket);
    assert_debug_snapshot!("focus_field_region", FocusField::Region);
    assert_debug_snapshot!("focus_field_prefix", FocusField::Prefix);
    assert_debug_snapshot!("focus_field_endpoint_url", FocusField::EndpointUrl);
    assert_debug_snapshot!("focus_field_access_key_id", FocusField::AccessKeyId);
    assert_debug_snapshot!("focus_field_secret_access_key", FocusField::SecretAccessKey);
    assert_debug_snapshot!("focus_field_path_style", FocusField::PathStyle);
}

#[test]
fn test_input_modes() {
    assert_debug_snapshot!("input_mode_normal", InputMode::Normal);
    assert_debug_snapshot!("input_mode_editing", InputMode::Editing);
}
