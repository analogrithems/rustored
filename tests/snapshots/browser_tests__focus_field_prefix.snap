---
source: tests/browser_tests.rs
expression: "FocusField::Prefix"
---
Prefix
